distributed/scheduler.py

```diff
@@ -3091,15 +3092,32 @@ def get_task_stream(self, comm=None, start=None, stop=None, count=None):
 
     @gen.coroutine
     def register_worker_callbacks(self, comm, setup=None):
-        """ Registers a setup function, and call it on every worker """
-        if setup is None:
-            raise gen.Return({})
+        """ Registers a set of event driven callback functions on workers for the given name.
+
+            setup must be a tuple of (name, serialized_function)
+        """
+        responses = {}
+
+        if setup is not None:
+            name, func = setup
 
-        self.worker_setups.append(setup)
+            # add the setup function to the list to run them on new clients.
+            self.worker_setups[name] = func
+
+            # trigger the setup function on the existing clients.
+            responses.update((yield self.broadcast(msg=dict(op='run', function=func))))
 
-        responses = yield self.broadcast(msg=dict(op='run', function=setup))
         raise gen.Return(responses)
 
+    @gen.coroutine
+    def unregister_worker_callbacks(self, comm, setup=None):
+        # add the setup function to the list to run them on new clients.
+        if setup is not None:
+            name, func = setup
+            self.worker_setups.pop(name)
+
+        raise gen.Return(None)
```

Comment:
This doesn't need to be a coroutine.  I recommend removing the `@gen.coroutine` line and replacing `raise gen.Return(...)` with `return ...`

Reply:
To be clear, the reason here is that this method never `yield`s, so it doesn't need to be a coroutine